pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podstore
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	DeleteQueueName = "pod_delete"
+	PodIPIndex      = "PodIP"
+	DelayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	curPod       cache.Indexer
+	prevPod      cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+}
+
+// newPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func newPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPod:       podInformer.GetIndexer(),
+		prevPod:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			DeleteFunc: s.onPodDelete,
+		})
+	s.curPod.AddIndexers(cache.Indexers{PodIPIndex: podIPIndexFunc})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return newPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing event delete")
+			return
+		}
+	}
+	if pod.GetDeletionTimestamp() == nil {
+		timeNow := time.Now()
+		pod.DeletionTimestamp = &metav1.Time{Time: timeNow}
```

Comment:
I believe event handlers should _not_ modify the object.

> ResourceEventHandler can handle notifications for events that happen to a resource. The events are informational only, so you can't return an error. The handlers MUST NOT modify the objects received; this concerns not only the top level of structure but all the data structures reachable from it.

source: https://pkg.go.dev/k8s.io/client-go/tools/cache#ResourceEventHandler

Reply:
Fixed. Thanks!